# Cancer Prediction Project

CRISP - ML Q methodology

1. a) Business problem Understanding

Cancer is a deadly disease and early detection can reduce fatality. For this we are creating an ML model to do the prediction based on symptoms and various aspects. So, if we can input the data model will predict the chance of cancer and we can inform to take necessary action

1. b) Data understanding:

We have data of people with their medical aspects like Tumor size, tumor grade, their symptoms and other aspects and if they have cancer or not.

In [1]:
import pandas as pd
df=pd.read_csv(r'C:\Users\punee\OneDrive\Desktop\python\Machinelearning\datapreprocessing\cancer_prediction_data.csv')
print(df.columns)
df.head(5)

Index(['Age', 'Gender', 'Tumor_Size', 'Tumor_Grade', 'Symptoms_Severity',
       'Family_History', 'Smoking_History', 'Alcohol_Consumption',
       'Exercise_Frequency', 'Cancer_Present'],
      dtype='object')


Age  Gender  Tumor_Size Tumor_Grade Symptoms_Severity Family_History  \
0  62.0    Male    9.763242        High              Mild            Yes   
1  33.0    Male    7.946532         NaN          Moderate            Yes   
2  33.0  Female         NaN        High          Moderate             No   
3  58.0  Female         NaN         Low              Mild            NaN   
4   NaN  Female    1.384231      Medium              Mild             No   

  Smoking_History Alcohol_Consumption Exercise_Frequency  Cancer_Present  
0   Former Smoker            Moderate                NaN               0  
1  Current Smoker                 NaN          Regularly               1  
2  Current Smoker                High             Rarely               1  
3  Current Smoker                 NaN             Rarely               1  
4   Former Smoker            Moderate       Occasionally               1

In [2]:
df.Gender.unique()

array(['Male', 'Female', nan], dtype=object)

In [3]:
df.Tumor_Grade.unique()

array(['High', nan, 'Low', 'Medium'], dtype=object)

In [4]:
df.Symptoms_Severity.unique()

array(['Mild', 'Moderate', 'Severe', nan], dtype=object)

In [5]:
df.Smoking_History.unique()

array(['Former Smoker', 'Current Smoker', 'Non-Smoker', nan], dtype=object)

In [6]:
df.Tumor_Grade.unique()

array(['High', nan, 'Low', 'Medium'], dtype=object)

In [7]:
df.Alcohol_Consumption.unique()

array(['Moderate', nan, 'High', 'Low'], dtype=object)

In [8]:
df.Exercise_Frequency.unique()

array([nan, 'Regularly', 'Rarely', 'Occasionally', 'Never'], dtype=object)

In [9]:
df.Cancer_Present.unique()

array([0, 1], dtype=int64)

## Data Attributes

Age: Numerical value representing the age of the individual.

Gender: Categorical value indicating the gender of the individual (Male/Female).

Tumor Size: Numerical value indicating the size of the tumor in centimeters.

Tumor Grade: Categorical value indicating the grade of the tumor (High, Medium, Low, NaN for missing data).

Symptoms Severity: Categorical value representing the severity of symptoms (Mild, Moderate).

Family History: Binary value indicating whether there is a family history of cancer (Yes/No).

Smoking History: Categorical value indicating the smoking history of the individual (Current Smoker, Former Smoker,Non-Smoker,nan).

Alcohol Consumption: Categorical value indicating the level of alcohol consumption (High, Moderate, low, NaN for missing data).

Exercise Frequency: Categorical value representing the frequency of exercise (Regularly, Rarely, Occasionally,Never, NaN for missing data).

Cancer Present: Binary value indicating whether cancer is present (1) or not (0).

### 2. Data Preparation

2. a) Exploratory Data analysis (EDA)

In [10]:
df

Age  Gender  Tumor_Size Tumor_Grade Symptoms_Severity Family_History  \
0     62.0    Male    9.763242        High              Mild            Yes   
1     33.0    Male    7.946532         NaN          Moderate            Yes   
2     33.0  Female         NaN        High          Moderate             No   
3     58.0  Female         NaN         Low              Mild            NaN   
4      NaN  Female    1.384231      Medium              Mild             No   
...    ...     ...         ...         ...               ...            ...   
1995  53.0     NaN    4.058134        High              Mild             No   
1996  57.0  Female    5.542532        High          Moderate            Yes   
1997   NaN  Female    0.866000      Medium              Mild            NaN   
1998  47.0     NaN    7.859645         Low            Severe            Yes   
1999  59.0    Male    5.109341      Medium              Mild            Yes   

     Smoking_History Alcohol_Consumption Exercise_Frequency  Cancer_Present  
0      Former Smoker            Moderate                NaN               0  
1     Current Smoker                 NaN          Regularly               1  
2     Current Smoker                High             Rarely               1  
3     Current Smoker                 NaN             Rarely               1  
4      Former Smoker            Moderate       Occasionally               1  
...              ...                 ...                ...             ...  
1995  Current Smoker                 Low                NaN               0  
1996   Former Smoker                High       Occasionally               1  
1997      Non-Smoker                High              Never               0  
1998   Former Smoker                 NaN             Rarely               1  
1999  Current Smoker                 Low       Occasionally               1  

[2000 rows x 10 columns]

In [11]:
df.isna().sum()

Age                    189
Gender                 206
Tumor_Size             171
Tumor_Grade            193
Symptoms_Severity      185
Family_History         208
Smoking_History        179
Alcohol_Consumption    679
Exercise_Frequency     195
Cancer_Present           0
dtype: int64

In [12]:
df.Age.count

<bound method Series.count of 0       62.0
1       33.0
2       33.0
3       58.0
4        NaN
        ... 
1995    53.0
1996    57.0
1997     NaN
1998    47.0
1999    59.0
Name: Age, Length: 2000, dtype: float64>

In [13]:
import numpy as np

df[df['Age'].isna()]

Age  Gender  Tumor_Size Tumor_Grade Symptoms_Severity Family_History  \
4     NaN  Female    1.384231      Medium              Mild             No   
34    NaN    Male    8.018982        High              Mild            Yes   
37    NaN    Male    9.842529      Medium          Moderate             No   
59    NaN    Male    6.980735      Medium            Severe             No   
62    NaN  Female    7.005259         NaN          Moderate            Yes   
...   ...     ...         ...         ...               ...            ...   
1969  NaN     NaN    0.782925         Low          Moderate            Yes   
1971  NaN  Female    7.286332         NaN            Severe             No   
1972  NaN  Female    7.747756        High              Mild            Yes   
1983  NaN    Male         NaN         Low              Mild            Yes   
1997  NaN  Female    0.866000      Medium              Mild            NaN   

     Smoking_History Alcohol_Consumption Exercise_Frequency  Cancer_Present  
4      Former Smoker            Moderate       Occasionally               1  
34    Current Smoker                High             Rarely               1  
37        Non-Smoker                High             Rarely               1  
59    Current Smoker                 NaN          Regularly               0  
62    Current Smoker            Moderate                NaN               1  
...              ...                 ...                ...             ...  
1969      Non-Smoker                High             Rarely               1  
1971  Current Smoker            Moderate             Rarely               0  
1972  Current Smoker                 NaN       Occasionally               0  
1983   Former Smoker            Moderate          Regularly               1  
1997      Non-Smoker                High              Never               0  

[189 rows x 10 columns]

here out of 2000, 189 are null for age. So not dropping the column

In [14]:
df

Age  Gender  Tumor_Size Tumor_Grade Symptoms_Severity Family_History  \
0     62.0    Male    9.763242        High              Mild            Yes   
1     33.0    Male    7.946532         NaN          Moderate            Yes   
2     33.0  Female         NaN        High          Moderate             No   
3     58.0  Female         NaN         Low              Mild            NaN   
4      NaN  Female    1.384231      Medium              Mild             No   
...    ...     ...         ...         ...               ...            ...   
1995  53.0     NaN    4.058134        High              Mild             No   
1996  57.0  Female    5.542532        High          Moderate            Yes   
1997   NaN  Female    0.866000      Medium              Mild            NaN   
1998  47.0     NaN    7.859645         Low            Severe            Yes   
1999  59.0    Male    5.109341      Medium              Mild            Yes   

     Smoking_History Alcohol_Consumption Exercise_Frequency  Cancer_Present  
0      Former Smoker            Moderate                NaN               0  
1     Current Smoker                 NaN          Regularly               1  
2     Current Smoker                High             Rarely               1  
3     Current Smoker                 NaN             Rarely               1  
4      Former Smoker            Moderate       Occasionally               1  
...              ...                 ...                ...             ...  
1995  Current Smoker                 Low                NaN               0  
1996   Former Smoker                High       Occasionally               1  
1997      Non-Smoker                High              Never               0  
1998   Former Smoker                 NaN             Rarely               1  
1999  Current Smoker                 Low       Occasionally               1  

[2000 rows x 10 columns]

In [15]:
## types of columns in the data

numeric=['Age','Tumor_Size']
ordinal=['Tumor_Grade','Symptoms_Severity','Alcohol_Consumption','Exercise_Frequency']
nominal=['Smoking_History','Gender','Family_History']

In [16]:
df.columns

Index(['Age', 'Gender', 'Tumor_Size', 'Tumor_Grade', 'Symptoms_Severity',
       'Family_History', 'Smoking_History', 'Alcohol_Consumption',
       'Exercise_Frequency', 'Cancer_Present'],
      dtype='object')

In [17]:
df[df['Cancer_Present']==1].isna().sum()

Age                     89
Gender                 103
Tumor_Size              87
Tumor_Grade             84
Symptoms_Severity       98
Family_History          91
Smoking_History         73
Alcohol_Consumption    357
Exercise_Frequency      84
Cancer_Present           0
dtype: int64

we can see here majority of cancer patients dont have their details properly mentioned and those are missing, we cannot drop columns

In [18]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler,OrdinalEncoder,OneHotEncoder

In [19]:
numeric_preprocess=Pipeline([('Mean Imputation',SimpleImputer(strategy='mean')),('Scaling',StandardScaler())])
numeric_preprocess

Pipeline(steps=[('Mean Imputation', SimpleImputer()),
                ('Scaling', StandardScaler())])

In [20]:
ordinal_preprocess=Pipeline([('Mode Imputation',SimpleImputer(strategy='most_frequent')),('Encoding',OrdinalEncoder())])
ordinal_preprocess

Pipeline(steps=[('Mode Imputation', SimpleImputer(strategy='most_frequent')),
                ('Encoding', OrdinalEncoder())])

In [21]:
nominal_preprocess=Pipeline([('Model Imputation',SimpleImputer(strategy='most_frequent')),('Encoding',OneHotEncoder())])
nominal_preprocess                      

Pipeline(steps=[('Model Imputation', SimpleImputer(strategy='most_frequent')),
                ('Encoding', OneHotEncoder())])

### Preprocessing

In [22]:
preprocess=ColumnTransformer([('numerical transformation',numeric_preprocess,numeric),('ordinal transformation',ordinal_preprocess,ordinal),('nominal transformation',nominal_preprocess,nominal)])
preprocess

ColumnTransformer(transformers=[('numerical transformation',
                                 Pipeline(steps=[('Mean Imputation',
                                                  SimpleImputer()),
                                                 ('Scaling',
                                                  StandardScaler())]),
                                 ['Age', 'Tumor_Size']),
                                ('ordinal transformation',
                                 Pipeline(steps=[('Mode Imputation',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('Encoding',
                                                  OrdinalEncoder())]),
                                 ['Tumor_Grade', 'Symptoms_Severity',
                                  'Alcohol_Consumption',
                                  'Exercise_Frequency']),
                                ('nominal transformation',
                                 Pipeline(steps=[('Model Imputation',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('Encoding',
                                                  OneHotEncoder())]),
                                 ['Smoking_History', 'Gender',
                                  'Family_History'])])

### 3. Model Building

In [23]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier

In [24]:
X=df.drop(columns='Cancer_Present',axis=1)

In [25]:
y=df['Cancer_Present']

In [26]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=20)

#### Decision tree

In [27]:
dt=Pipeline([('Data preprocessing',preprocess),('Algorithm',DecisionTreeClassifier())])
dt.fit(X_train,y_train).score(X_test,y_test)

0.47

#### Logistic Regression

In [28]:
log=Pipeline([('Data Preprocessing',preprocess),('Algorithm',LogisticRegression())])
log.fit(X_train,y_train).score(X_test,y_test)

0.535

#### KNN

In [29]:
kn=Pipeline([('Data Preprocessing',preprocess),('Algorithm',KNeighborsClassifier())])
kn.fit(X_train,y_train).score(X_test,y_test)

0.52

#### SVM

In [30]:
svm=Pipeline([('Data Preprocessing',preprocess),('Algorithm',SVC())])
svm.fit(X_train,y_train).score(X_test,y_test)

0.48

#### XG boost

In [31]:
from xgboost import XGBClassifier

In [32]:
xg=Pipeline([('Data Preprocessing',preprocess),('Algorithm',XGBClassifier())])
xg.fit(X_train,y_train).score(X_test,y_test)

0.4825

#### Random Forest

In [33]:
from sklearn.ensemble import RandomForestClassifier

rf=Pipeline([('Data Preprocessing',preprocess),('Algorithm',RandomForestClassifier())]) 
rf.fit(X_train,y_train).score(X_test,y_test)

0.5

#### Model Deployment

In [34]:
import pickle
pickle.dump(dt,open('dt.pkl','wb'))
pickle.dump(log,open('log.pkl','wb'))
pickle.dump(kn,open('kn.pkl','wb'))
pickle.dump(svm,open('svm.pkl','wb'))
pickle.dump(xg,open('xg.pkl','wb'))
pickle.dump(rf,open('rf.pkl','wb'))

In [35]:
df

Age  Gender  Tumor_Size Tumor_Grade Symptoms_Severity Family_History  \
0     62.0    Male    9.763242        High              Mild            Yes   
1     33.0    Male    7.946532         NaN          Moderate            Yes   
2     33.0  Female         NaN        High          Moderate             No   
3     58.0  Female         NaN         Low              Mild            NaN   
4      NaN  Female    1.384231      Medium              Mild             No   
...    ...     ...         ...         ...               ...            ...   
1995  53.0     NaN    4.058134        High              Mild             No   
1996  57.0  Female    5.542532        High          Moderate            Yes   
1997   NaN  Female    0.866000      Medium              Mild            NaN   
1998  47.0     NaN    7.859645         Low            Severe            Yes   
1999  59.0    Male    5.109341      Medium              Mild            Yes   

     Smoking_History Alcohol_Consumption Exercise_Frequency  Cancer_Present  
0      Former Smoker            Moderate                NaN               0  
1     Current Smoker                 NaN          Regularly               1  
2     Current Smoker                High             Rarely               1  
3     Current Smoker                 NaN             Rarely               1  
4      Former Smoker            Moderate       Occasionally               1  
...              ...                 ...                ...             ...  
1995  Current Smoker                 Low                NaN               0  
1996   Former Smoker                High       Occasionally               1  
1997      Non-Smoker                High              Never               0  
1998   Former Smoker                 NaN             Rarely               1  
1999  Current Smoker                 Low       Occasionally               1  

[2000 rows x 10 columns]

In [37]:
df.describe(include='all')

Age  Gender   Tumor_Size Tumor_Grade Symptoms_Severity  \
count   1811.000000    1794  1829.000000        1807              1815   
unique          NaN       2          NaN           3                 3   
top             NaN  Female          NaN         Low          Moderate   
freq            NaN     930          NaN         619               614   
mean      49.267808     NaN     5.133781         NaN               NaN   
std       17.361243     NaN     2.781335         NaN               NaN   
min       20.000000     NaN     0.507471         NaN               NaN   
25%       35.000000     NaN     2.700941         NaN               NaN   
50%       49.000000     NaN     5.079521         NaN               NaN   
75%       64.000000     NaN     7.588271         NaN               NaN   
max       79.000000     NaN     9.998166         NaN               NaN   

       Family_History Smoking_History Alcohol_Consumption Exercise_Frequency  \
count            1792            1821                1321               1805   
unique              2               3                   3                  4   
top               Yes   Former Smoker                High              Never   
freq              897             624                 446                467   
mean              NaN             NaN                 NaN                NaN   
std               NaN             NaN                 NaN                NaN   
min               NaN             NaN                 NaN                NaN   
25%               NaN             NaN                 NaN                NaN   
50%               NaN             NaN                 NaN                NaN   
75%               NaN             NaN                 NaN                NaN   
max               NaN             NaN                 NaN                NaN   

        Cancer_Present  
count      2000.000000  
unique             NaN  
top                NaN  
freq               NaN  
mean          0.491500  
std           0.500053  
min           0.000000  
25%           0.000000  
50%           0.000000  
75%           1.000000  
max           1.000000

In [38]:
df[ordinal]

Tumor_Grade Symptoms_Severity Alcohol_Consumption Exercise_Frequency
0           High              Mild            Moderate                NaN
1            NaN          Moderate                 NaN          Regularly
2           High          Moderate                High             Rarely
3            Low              Mild                 NaN             Rarely
4         Medium              Mild            Moderate       Occasionally
...          ...               ...                 ...                ...
1995        High              Mild                 Low                NaN
1996        High          Moderate                High       Occasionally
1997      Medium              Mild                High              Never
1998         Low            Severe                 NaN             Rarely
1999      Medium              Mild                 Low       Occasionally

[2000 rows x 4 columns]

In [40]:
df[ordinal].value_counts()

Tumor_Grade  Symptoms_Severity  Alcohol_Consumption  Exercise_Frequency
High         Mild               High                 Rarely                19
             Severe             Moderate             Occasionally          16
             Moderate           Moderate             Never                 14
Low          Moderate           Low                  Regularly             14
Medium       Mild               High                 Occasionally          13
                                                                           ..
             Severe             Moderate             Never                  4
High         Mild               Low                  Never                  3
             Severe             High                 Never                  3
             Moderate           Low                  Regularly              3
Medium       Mild               Moderate             Regularly              3
Name: count, Length: 108, dtype: int64

In [42]:
df.columns

Index(['Age', 'Gender', 'Tumor_Size', 'Tumor_Grade', 'Symptoms_Severity',
       'Family_History', 'Smoking_History', 'Alcohol_Consumption',
       'Exercise_Frequency', 'Cancer_Present'],
      dtype='object')

In [43]:
df.Tumor_Grade.unique()

array(['High', nan, 'Low', 'Medium'], dtype=object)

In [45]:
df.Symptoms_Severity.unique()

array(['Mild', 'Moderate', 'Severe', nan], dtype=object)

In [46]:
df.Exercise_Frequency.unique()

array([nan, 'Regularly', 'Rarely', 'Occasionally', 'Never'], dtype=object)

In [47]:
df.columns

Index(['Age', 'Gender', 'Tumor_Size', 'Tumor_Grade', 'Symptoms_Severity',
       'Family_History', 'Smoking_History', 'Alcohol_Consumption',
       'Exercise_Frequency', 'Cancer_Present'],
      dtype='object')

In [48]:
df.Smoking_History.unique()

array(['Former Smoker', 'Current Smoker', 'Non-Smoker', nan], dtype=object)

In [49]:
df.Alcohol_Consumption.unique()

array(['Moderate', nan, 'High', 'Low'], dtype=object)

In [50]:
a=0.25

In [51]:
import seaborn as sns

TypeError: object of type 'float' has no len()